rustls/src/client/tls12.rs

```diff
@@ -40,29 +40,45 @@ mod server_hello {
     use super::*;
     use crate::client::hs::{ClientHelloInput, ClientSessionValue};
     use crate::msgs::handshake::ServerHelloPayload;
+    use crate::sealed::Sealed;
 
-    pub(in crate::client) struct CompleteServerHelloHandling {
-        pub(in crate::client) randoms: ConnectionRandoms,
-        pub(in crate::client) transcript: HandshakeHash,
-        pub(in crate::client) input: ClientHelloInput,
-    }
+    pub(crate) static TLS12_HANDLER: &dyn Tls12Handler = &Handler;
 
-    impl CompleteServerHelloHandling {
-        pub(in crate::client) fn handle_server_hello(
-            mut self,
+    pub(crate) trait Tls12Handler: Sealed + Send + Sync + core::fmt::Debug {
+        fn handle_server_hello(
+            &self,
             cx: &mut ClientContext<'_>,
+            server_hello: &ServerHelloPayload,
+            randoms: ConnectionRandoms,
             suite: &'static Tls12CipherSuite,
+            transcript: HandshakeHash,
+            tls13_supported: bool,
+            input: ClientHelloInput,
+        ) -> hs::NextStateOrError<'static>;
+    }
+
+    #[derive(Debug)]
+    struct Handler;
```

Comment:
Could this be replaced by an existing type? Maybe `Tls12Version` or `Tls12CipherSuite`?

Reply:
So we need a value of the type to take a dyn reference. For `Tls12Version` that value would contain itself. For `Tls12CipherSuite` that would need to refer to itself too, via the `protocol_version` field.

The other way of structuring this is to use function pointers. That is a bit less idiomatic rust and does less well at hiding the types of arguments (but can be made to work).